routes/eth.go

```diff
@@ -367,14 +369,104 @@ func (fes *APIServer) QueryETHRPC(ww http.ResponseWriter, req *http.Request) {
 	}
 }
 
+type MetamaskSignInRequest struct {
+	RecipientPublicKey  string
+	RecipientEthAddress string
+	AmountNanos         uint64
+	Signer              []byte
+	Message             []byte
+	Signature           []byte
+}
+type MetamaskSignInResponse struct {
+	TxnHash *lib.BlockHash
+}
+
+func (fes *APIServer) MetamaskSignIn(ww http.ResponseWriter, req *http.Request) {
+	// Give the user starter deso if this is their first time signing in with through metamask and if they don't have Deso
+	decoder := json.NewDecoder(io.LimitReader(req.Body, MaxRequestBodySizeBytes))
+
+	// Validate the  request object
+	requestData := MetamaskSignInRequest{}
+	if err := decoder.Decode(&requestData); err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("MetamaskSignin: Problem parsing request body: %v", err))
+		return
+	}
+	recipientBytePK := []byte(requestData.RecipientPublicKey)
+
+	// Validate that the user doesn't have Deso already
+	desoBalance, desoBalanceErr := fes.getBalanceForPubKey(recipientBytePK)
+	// balance check TESTED
+	if desoBalance != 0 {
+		_AddBadRequestError(ww, fmt.Sprintf("MetamaskSignin: Deso balance greater than zero", desoBalanceErr))
+		return
+	}
+
+	hasReceivedAirdrop, err := fes.GlobalState.Get(GlobalStateKeyMetamaskAirdrop(recipientBytePK))
+	// check if they have recieved the airdrop TESTED TODO switch 2 back to 1
+	if bytes.Equal(hasReceivedAirdrop, []byte{1}) {
+		_AddBadRequestError(ww, fmt.Sprintf("MetamaskSignin: Account has already received airdrop"))
+		return
+	}
+	if err = fes.GlobalState.Put(GlobalStateKeyMetamaskAirdrop(recipientBytePK), []byte{1}); err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("MetamaskSignin: Account failed to update airdrop state"))
+		return
+	}
+
+	// validate the user's eth balance
+	params := []interface{}{requestData.RecipientEthAddress, "latest"}
+	infuraResponse, err := fes.ExecuteETHRPCRequest("eth_getBalance", params)
+	if infuraResponse == nil || err != nil {
+		_AddBadRequestError(ww, fmt.Sprintf("something wrong with fetching eth balance", err))
+		return
+	}
+	ethBalance := strings.Split(infuraResponse.Result.(string), "x")[1]
+	ethBalanceInt, _ := strconv.ParseInt(ethBalance, 16, 64)
+	ethBalanceInt64 := ethBalanceInt / 10000000000000000
```

Comment:
Divisions are scary. Maybe we should just stick with big numbers. So like instead of the ethBalanceInt64 < 100, we just do int64(1000000000000000000)